File: rust-substrate/crates/apeireth-core/src/deliberation.rs
//! Deliberation Engine — Rust implementation of Thought Branching (Tree-of-Thoughts).
//!
//! 主人 20:29 真哲学指令: "底层记得用rust, 我们追求极致"
//! "思考为核, ASI绝对会自己思考, 任何LLM接入即ASI"
//!
//! Python `apeireth/deliberation.py` 是高层 wrapper.
//! 这个 Rust 实现是 hot path (ToT BFS/DFS 树搜索) — 极致性能.

use serde::{Deserialize, Serialize};

/// One thought step in deliberation.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ThoughtStep {
    pub step_id: String,
    pub step_type: StepType,
    pub content: String,
    pub confidence: f64,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
pub enum StepType {
    Thought,
    Action,
    Observation,
    Critique,
}

/// A branch in Tree-of-Thoughts — explores one hypothesis.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ThoughtBranch {
    pub branch_id: String,
    pub hypothesis: String,
    pub steps: Vec<ThoughtStep>,
    pub score: f64,
    pub status: BranchStatus,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
pub enum BranchStatus {
    Active,
    Completed,
    Pruned,
    Selected,
}

/// Result of deliberation — best plan + reasoning summary.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DeliberationResult {
    pub deliberation_id: String,
    pub query: String,
    pub mode: String,
    pub branches: Vec<ThoughtBranch>,
    pub selected_branch_id: Option<String>,
    pub final_plan: Vec<String>,
    pub self_score: f64,
    pub total_steps: usize,
}

/// Tree-of-Thoughts engine — Rust hot path for ASI deliberation.
pub struct TotEngine {
    max_branches: usize,
    max_depth: usize,
}

impl TotEngine {
    pub fn new(max_branches: usize, max_depth: usize) -> Self {
        Self { max_branches, max_depth }
    }

    /// Score a branch by averaging its steps' confidences.
    pub fn score(branch: &ThoughtBranch) -> f64 {
        if branch.steps.is_empty() {
            return 0.0;
        }
        let sum: f64 = branch.steps.iter().map(|s| s.confidence).sum();
        sum / branch.steps.len() as f64
    }

    /// Select the best branch (highest score).
    pub fn select_best(branches: &[ThoughtBranch]) -> Option<usize> {
        if branches.is_empty() {
            return None;
        }
        let mut best_idx = 0;
        let mut best_score = Self::score(&branches[0]);
        for (i, b) in branches.iter().enumerate().skip(1) {
            let s = Self::score(b);
            if s > best_score {
                best_score = s;
                best_idx = i;
            }
        }
        Some(best_idx)
    }

    /// BFS-style expansion: explore N hypotheses in parallel.
    /// (real LLM calls happen in Python layer, this is the structural search)
    pub fn select_top_k(branches: &[ThoughtBranch], k: usize) -> Vec<usize> {
        let mut indexed: Vec<(usize, f64)> = branches
            .iter()
            .enumerate()
            .map(|(i, b)| (i, Self::score(b)))
            .collect();
        indexed.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));
        indexed.into_iter().take(k).map(|(i, _)| i).collect()
    }
}

/// Default ToT config for ASI (主人 20:29 "极致").
pub fn default_asi_tot_engine() -> TotEngine {
    TotEngine::new(5, 8)  // 5 branches, depth 8
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_score_branch() {
        let branch = ThoughtBranch {
            branch_id: "b1".into(),
            hypothesis: "test".into(),
            steps: vec![
                ThoughtStep { step_id: "s1".into(), step_type: StepType::Thought, content: "a".into(), confidence: 0.8 },
                ThoughtStep { step_id: "s2".into(), step_type: StepType::Action, content: "b".into(), confidence: 0.6 },
            ],
            score: 0.0,
            status: BranchStatus::Active,
        };
        assert!((TotEngine::score(&branch) - 0.7).abs() < 1e-9);
    }

    #[test]
    fn test_select_best() {
        let branches = vec![
            ThoughtBranch {
                branch_id: "b1".into(),
                hypothesis: "h1".into(),
                steps: vec![ThoughtStep { step_id: "s1".into(), step_type: StepType::Thought, content: "a".into(), confidence: 0.3 }],
                score: 0.0,
                status: BranchStatus::Active,
            },
            ThoughtBranch {
                branch_id: "b2".into(),
                hypothesis: "h2".into(),
                steps: vec![ThoughtStep { step_id: "s2".into(), step_type: StepType::Thought, content: "b".into(), confidence: 0.9 }],
                score: 0.0,
                status: BranchStatus::Active,
            },
        ];
        assert_eq!(TotEngine::select_best(&branches), Some(1));
    }

    #[test]
    fn test_top_k() {
        let branches = vec![
            ThoughtBranch { branch_id: "a".into(), hypothesis: "".into(), steps: vec![ThoughtStep { step_id: "".into(), step_type: StepType::Thought, content: "".into(), confidence: 0.1 }], score: 0.0, status: BranchStatus::Active },
            ThoughtBranch { branch_id: "b".into(), hypothesis: "".into(), steps: vec![ThoughtStep { step_id: "".into(), step_type: StepType::Thought, content: "".into(), confidence: 0.5 }], score: 0.0, status: BranchStatus::Active },
            ThoughtBranch { branch_id: "c".into(), hypothesis: "".into(), steps: vec![ThoughtStep { step_id: "".into(), step_type: StepType::Thought, content: "".into(), confidence: 0.3 }], score: 0.0, status: BranchStatus::Active },
        ];
        let top = TotEngine::select_top_k(&branches, 2);
        assert_eq!(top, vec![1, 2]);  // b (0.5), c (0.3)
    }

    #[test]
    fn test_default_asi_tot_engine() {
        let engine = default_asi_tot_engine();
        assert_eq!(engine.max_branches, 5);
        assert_eq!(engine.max_depth, 8);
    }
}

File: apeireth/deliberation.py
"""Phase 19 DeliberationEngine — ASI 思考层 (Thinking Layer).

主人 20:29 哲学指令 (历史性):
  "底层记得用rust，我们追求极致"
  "除了记忆，思考也要重视"
  "ASI绝对是会自己思考的"
  "我们做的这个ASI基座也要无限逼近"
  "ASI就是我们的目标，让任何大模型接入我们的平台后成为ASI"

5 大真生产参考:
  1. DeepSeek-R1 (arxiv 2501.12948) — chain-of-thought + RL
  2. Tree of Thoughts (arxiv 2305.10601) — 多路径探索
  3. ReAct (arxiv 2210.03629) — reasoning + acting interleaved
  4. Self-Refine (arxiv 2310.01798) — 自我反馈 + 修正
  5. Reflexion (arxiv 2303.17651) — verbal RL

Karpathy 准则应用:
  1. Think Before Coding: 真思考 = 多路径探索 + 自我批评
  2. Simplicity First: v0.1 只 3 模式 (linear / tot / reflexion)
  3. Surgical Changes: 不改已有模块, 只加 thinking 入口
  4. Goal-Driven Execution: verifiable = 思考后给出 plan + 自我评分
"""
from __future__ import annotations

import time
import uuid
from dataclasses import dataclass, field, asdict
from typing import Optional


DELIBERATION_VERSION = "0.1.0"


@dataclass
class ThoughtStep:
    """One step in a deliberation chain."""
    step_id: str
    step_type: str            # "thought" | "action" | "observation" | "critique"
    content: str
    confidence: float          # 0-1, self-rated
    parent_step_id: Optional[str] = None
    ts: float = field(default_factory=time.time)

    def to_dict(self) -> dict:
        return asdict(self)


@dataclass
class ThoughtBranch:
    """A branch in Tree of Thoughts — explores one hypothesis."""
    branch_id: str
    hypothesis: str           # what we're testing
    steps: list[ThoughtStep] = field(default_factory=list)
    score: float = 0.0
    status: str = "active"     # active | completed | pruned | selected

    def to_dict(self) -> dict:
        return asdict(self)


@dataclass
class DeliberationResult:
    """Final output of deliberation — plan + reasoning chain + self-rating."""
    deliberation_id: str
    query: str
    mode: str                 # "linear" | "tot" | "reflexion"
    branches: list[ThoughtBranch] = field(default_factory=list)
    selected_branch_id: Optional[str] = None
    final_plan: list[str] = field(default_factory=list)
    self_score: float = 0.0
    reasoning_summary: str = ""
    total_steps: int = 0
    ts: float = field(default_factory=time.time)

    def to_dict(self) -> dict:
        return asdict(self)


class DeliberationEngine:
    """Central AI 思考引擎 — LLM-agnostic pluggable.

    3 真生产模式:
      - linear: chain-of-thought (DeepSeek-R1 借鉴)
      - tot:    tree-of-thought (Yao 2023 借鉴, 多路径探索)
      - reflexion: self-feedback + refine (Shinn 2023 借鉴)

    LLM-agnostic: 任何 LLM (Claude / DeepSeek / Qwen / GPT) 通过 call_llm(prompt) 接入.
    """

    def __init__(self, call_llm=None, max_steps: int = 6):
        """call_llm(prompt: str) -> str — any LLM function."""
        self.call_llm = call_llm or self._default_llm
        self.max_steps = max_steps
        self.history: list[DeliberationResult] = []

    def _default_llm(self, prompt: str) -> str:
        """Default no-op LLM — returns template response (for testing without LLM)."""
        return f"[template-LLM] {prompt[:100]}... -> [simulated-response]"

    def deliberate(self, query: str, mode: str = "linear", context: str = "") -> DeliberationResult:
        """Main entry — think about query in given mode."""
        result = DeliberationResult(
            deliberation_id=uuid.uuid4().hex[:12],
            query=query,
            mode=mode,
        )

        if mode == "linear":
            self._linear(query, context, result)
        elif mode == "tot":
            self._tot(query, context, result)
        elif mode == "reflexion":
            self._reflexion(query, context, result)
        else:
            raise ValueError(f"Unknown mode: {mode}")

        self.history.append(result)
        return result

    # ===== Mode 1: Linear (chain-of-thought, DeepSeek-R1 借鉴) =====

    def _linear(self, query: str, context: str, result: DeliberationResult) -> None:
        """Linear CoT — sequential thoughts + action + observation."""
        branch = ThoughtBranch(branch_id="linear_0", hypothesis=query)
        # Step 1: 思考 (Thought)
        s1 = ThoughtStep(
            step_id=uuid.uuid4().hex[:8],
            step_type="thought",
            content=self.call_llm(f"Think about: {query}\nContext: {context}\nStep 1: What is the core question?"),
            confidence=0.5,
        )
        branch.steps.append(s1)
        # Step 2: 行动 (Action)
        s2 = ThoughtStep(
            step_id=uuid.uuid4().hex[:8],
            step_type="action",
            content=self.call_llm(f"Given thought: {s1.content[:200]}\nWhat action to take?"),
            confidence=0.6,
            parent_step_id=s1.step_id,
        )
        branch.steps.append(s2)
        # Step 3: 观察 (Observation)
        s3 = ThoughtStep(
            step_id=uuid.uuid4().hex[:8],
            step_type="observation",
            content=self.call_llm(f"Action: {s2.content[:200]}\nWhat did we observe?"),
            confidence=0.7,
            parent_step_id=s2.step_id,
        )
        branch.steps.append(s3)
        # Step 4: 反思 (Critique)
        s4 = ThoughtStep(
            step_id=uuid.uuid4().hex[:8],
            step_type="critique",
            content=self.call_llm(f"Sequence: Thought→Action→Observation. Any errors?"),
            confidence=0.65,
            parent_step_id=s3.step_id,
        )
        branch.steps.append(s4)
        # Final
        branch.score = (s1.confidence + s2.confidence + s3.confidence + s4.confidence) / 4
        branch.status = "completed"
        result.branches.append(branch)
        result.selected_branch_id = branch.branch_id
        result.final_plan = [s.content[:100] for s in branch.steps]
        result.self_score = branch.score
        result.reasoning_summary = f"Linear CoT: {len(branch.steps)} steps, score={branch.score:.2f}"
        result.total_steps = len(branch.steps)

    # ===== Mode 2: Tree of Thoughts (多路径探索, Yao 2023 借鉴) =====

    def _tot(self, query: str, context: str, result: DeliberationResult) -> None:
        """ToT — generate 3 hypotheses, score each, select best."""
        hypotheses = [
            f"H1: Approach A — directly solve {query[:50]}",
            f"H2: Approach B — decompose {query[:50]} into subproblems",
            f"H3: Approach C — analogize {query[:50]} to known patterns",
        ]
        # Step 1: Generate 3 branches
        for hyp in hypotheses:
            branch = ThoughtBranch(
                branch_id=uuid.uuid4().hex[:8],
                hypothesis=hyp,
            )
            s = ThoughtStep(
                step_id=uuid.uuid4().hex[:8],
                step_type="thought",
                content=self.call_llm(f"For {hyp}\nContext: {context}\nGenerate next step"),
                confidence=0.5,
            )
            branch.steps.append(s)
            result.branches.append(branch)
        # Step 2: Score each branch
        for branch in result.branches:
            s = ThoughtStep(
                step_id=uuid.uuid4().hex[:8],
                step_type="critique",
                content=self.call_llm(f"Evaluate: {branch.hypothesis}\nLast step: {branch.steps[-1].content[:200]}\nRate feasibility"),
                confidence=0.5,
            )
            branch.steps.append(s)
            # score = avg of confidences
            branch.score = sum(x.confidence for x in branch.steps) / len(branch.steps)
        # Step 3: Select best branch
        best = max(result.branches, key=lambda b: b.score)
        best.status = "selected"
        for b in result.branches:
            if b.branch_id != best.branch_id:
                b.status = "pruned"
        result.selected_branch_id = best.branch_id
        result.final_plan = [s.content[:100] for s in best.steps]
        result.self_score = best.score
        result.reasoning_summary = f"ToT: {len(result.branches)} branches, best score={best.score:.2f}"
        result.total_steps = sum(len(b.steps) for b in result.branches)

    # ===== Mode 3: Reflexion (self-feedback, Shinn 2023 借鉴) =====

    def _reflexion(self, query: str, context: str, result: DeliberationResult) -> None:
        """Reflexion — initial answer → critique → refine → final."""
        branch = ThoughtBranch(branch_id="reflexion_0", hypothesis=query)
        # Step 1: Initial answer
        s1 = ThoughtStep(
            step_id=uuid.uuid4().hex[:8],
            step_type="thought",
            content=self.call_llm(f"Answer: {query}\nContext: {context}\nInitial answer:"),
            confidence=0.6,
        )
        branch.steps.append(s1)
        # Step 2: Self-critique
        s2 = ThoughtStep(
            step_id=uuid.uuid4().hex[:8],
            step_type="critique",
            content=self.call_llm(f"Initial: {s1.content[:200]}\nWhat's wrong with this? Self-critique:"),
            confidence=0.7,
            parent_step_id=s1.step_id,
        )
        branch.steps.append(s2)
        # Step 3: Refined answer
        s3 = ThoughtStep(
            step_id=uuid.uuid4().hex[:8],
            step_type="thought",
            content=self.call_llm(f"Initial: {s1.content[:200]}\nCritique: {s2.content[:200]}\nRefined:"),
            confidence=0.8,
            parent_step_id=s2.step_id,
        )
        branch.steps.append(s3)
        branch.score = (s1.confidence + s2.confidence + s3.confidence) / 3
        branch.status = "completed"
        result.branches.append(branch)
        result.selected_branch_id = branch.branch_id
        result.final_plan = [s.content[:100] for s in branch.steps]
        result.self_score = branch.score
        result.reasoning_summary = f"Reflexion: 3 steps (initial→critique→refined), score={branch.score:.2f}"
        result.total_steps = len(branch.steps)


def make_default_deliberation_engine(call_llm=None) -> DeliberationEngine:
    return DeliberationEngine(call_llm=call_llm)


__all__ = [
    "DELIBERATION_VERSION",
    "ThoughtStep",
    "ThoughtBranch",
    "DeliberationResult",
    "DeliberationEngine",
    "make_default_deliberation_engine",
]